internal/core/app_status_report.go

```diff
@@ -0,0 +1,221 @@
+package core
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/zclconf/go-cty/cty"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/go-argmapper"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl/v2"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	sdk "github.com/hashicorp/waypoint-plugin-sdk/proto/gen"
+	"github.com/hashicorp/waypoint/internal/config"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+// Builds a status report on the given deployment
+func (a *App) StatusReport(
+	ctx context.Context,
+	deployTarget *pb.Deployment,
+	releaseTarget *pb.Release,
+) (*pb.StatusReport, *sdk.StatusReport, error) {
+	// Query the artifact
+	var artifact *pb.PushedArtifact
+	if deployTarget != nil && deployTarget.Preload != nil &&
+		deployTarget.Preload.Artifact != nil {
+		artifact = deployTarget.Preload.Artifact
+	}
+	if artifact == nil && deployTarget != nil {
+		a.logger.Debug("querying artifact", "artifact_id", deployTarget.ArtifactId)
+		resp, err := a.client.GetPushedArtifact(ctx, &pb.GetPushedArtifactRequest{
+			Ref: &pb.Ref_Operation{
+				Target: &pb.Ref_Operation_Id{
+					Id: deployTarget.ArtifactId,
+				},
+			},
+		})
+		if err != nil {
+			a.logger.Error("error querying artifact",
+				"artifact_id", deployTarget.ArtifactId,
+				"error", err)
+			return nil, nil, err
+		}
+
+		artifact = resp
+	}
+
+	var evalCtx hcl.EvalContext
+	evalCtx.Variables = map[string]cty.Value{}
+	if err := evalCtxTemplateProto(&evalCtx, "artifact", artifact); err != nil {
+		a.logger.Warn("failed to prepare template variables, will not be available",
+			"err", err)
+	}
+
+	c, err := a.createStatusReporter(ctx, &evalCtx)
+	if status.Code(err) == codes.Unimplemented {
+		c = nil
+		err = nil
+	}
+	if err != nil {
+		a.logger.Error("error creating component in platform", "error", err)
+		return nil, nil, err
+	}
+	defer c.Close()
```

Comment:
Maybe I'm wrong, but if `c` is assigned nil on line 62 then is it possible to seg fault here if c is being de-refenced? (I didn't check to see if c is actually being derefenced, just thought I'd ask)